io/concatenateHDF5.m

```diff
@@ -0,0 +1,157 @@
+            numBadFrames = length(tif_info);
+            testFrames = testFrames + numBadFrames;
+        else
+            %%% Good trial
+            badTrial = 0;
+        end
+    end
+
+    if(badTrial == 0)
+        %%% Add the good trial to the hdf5 file
+        trialCount = trialCount+1;
+        tifName = fullfile(tifDir,list(i).name);
+        tifInfo = imfinfo(tifName);
+        tifFile = Tiff(tifName,'r');
+        oriFrames = length(tifInfo);
+        
+        xmlName = [tifName(1:end-3),'xml'];
+        xmlData = parse_miniscope_xml(xmlName);
+        
+        %%% Determine if frames were dropped by the miniscope during the
+        %%% trial
+        if(str2num(xmlData.dropped_count) ~= 0)
+            droppedFrames = str2num(xmlData.dropped);
+        else
+            numFrames = oriFrames;
+            droppedFrames = 0;
+        end
+        
+        %%% Read in the images, trim the frames at the beginning and end 
+        %%% of the trial (set by the 'trim' argument), and downsample by the 
+        %%% 'downsmpFactor' argument
+        imageStack = uint16(zeros(floor(rows*downsmpFactor),floor(cols*downsmpFactor),oriFrames-trim(1)-trim(2)));
+        if(downsmpFactor == 1)
+            for j=1:oriFrames-trim(1)-trim(2)
+                tifFile.setDirectory(j+trim(1));
+                imageStack(:,:,j) = uint16(tifFile.read());
+            end
+        else
+            for j=1:oriFrames-trim(1)-trim(2)
+                tifFile.setDirectory(j+trim(1));
+                imageStack(:,:,j) = uint16(imresize(tifFile.read(),downsmpFactor,'bilinear'));
+            end
+        end
+        
+        %%% Replace any dropped frames with the frame immediately
+        %%% preceeding it
+        if(droppedFrames ~=0)
+            for j=1:length(droppedFrames)
+                disp('DroppedFrame');
+                droppedFrame = droppedFrames(j);
+                if((droppedFrame > trim(1)) && (droppedFrame < oriFrames-trim(2)))
+                    droppedFrame = droppedFrame+trim(1);
+                    frontStack = imageStack(:,:,1:droppedFrame-1);
+                    frontStack = cat(3,frontStack,imageStack(:,:,droppedFrame-1));
+                    backStack = imageStack(:,:,droppedFrame:end);
+                    newImageStack = cat(3,frontStack,backStack);
+                    clear imageStack
+                    imageStack = newImageStack;
+                end
+            end
+        end
+        
+        numFrames = size(imageStack,3);
+        [dRows,dCols] = size(imageStack(:,:,1));
+        
+        h5write(fullfile(outputDir,hdf5Name),dataset,imageStack,[1,1,totalFrames],[dRows,dCols,numFrames]);
+        
+        %%%%%%%%%%%%%% NEED TO FIX %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+        %%% Currently does not take into account added frames if some were
+        %%% originally dropped - only tested on cases without dropped
+        %%% frames

```

Comment:
So we should assume that the original text file will have correct frames while the movie itself might have dropped frames so that after the frames are added to the movie, the current code will be correct (just want to be sure I understand).

----- Original Message -----
From: "Tony" <notifications@github.com>
To: "inanhkn/analysis" <analysis@noreply.github.com>
Cc: "jmaxey" <jmaxey@stanford.edu>
Sent: Monday, February 2, 2015 10:27:46 AM
Subject: Re: [analysis] Concatenate all frames from a trial into an hdf5 (#16)

> +                    newImageStack = cat(3,frontStack,backStack);
> +                    clear imageStack
> +                    imageStack = newImageStack;
> +                end
> +            end
> +        end
> +        
> +        numFrames = size(imageStack,3);
> +        [dRows,dCols] = size(imageStack(:,:,1));
> +        
> +        h5write(fullfile(outputDir,hdf5Name),dataset,imageStack,[1,1,totalFrames],[dRows,dCols,numFrames]);
> +        
> +        %%%%%%%%%%%%%% NEED TO FIX %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
> +        %%% Currently does not take into account added frames if some were
> +        %%% originally dropped - only tested on cases without dropped
> +        %%% frames

Let's operate with the assumption that the text file from the plus maze has the absolutely correct frame counts. (In cases where the text file and the Miniscope differ, we can manually edit the text file to correct the discrepancy.)

Also, `io/compress_frame_indices.m` already takes care of recomputing trial frame indices, given the original frame indices and the `trim` parameter.

---
Reply to this email directly or view it on GitHub:
https://github.com/inanhkn/analysis/pull/16/files#r23945563